types/ejs/index.d.ts

```diff
@@ -28,14 +28,22 @@ export function resolveInclude(name: string, filename: string, isDir: boolean):
 /**
  * Compile the given `str` of ejs into a template function.
  */
-export function compile(template: string, opts?: Options): (TemplateFunction);
+export function compile(template: string, opts?: Options & { async: false }): TemplateFunction;
+export function compile(template: string, opts: Options & { async: true }): AsyncTemplateFunction;
+export function compile(template: string, opts: Options & { async: boolean }): TemplateFunction | AsyncTemplateFunction;
+export function compile(template: string, opts: Exclude<Options, { async: any }>): TemplateFunction;
```

Comment:
Would it be correct to just replace the `Exclude` bit with `Pick<Options, Exclude<keyof Options, 'async'>>`, per [the 2.8 release post](http://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html)? If you ask me, the explicit conditional types are even harder to read than the string of overloads, and introducing a generic parameter that isn't actually meaningfully generic doesn't help either -- not to mention it breaks "object literal may only specify known properties" warnings.

~~(Really, `Omit` shouldn't have been dropped from lib at all, and `Exclude` should be better-documented, but that's not for here)~~

The last case is intended to be the "true" signature -- the one that would carry the implementation body if it was here. Is it not required for declarations? I'll remove it if so.

Reply:
After looking at this a bit more, here's another approach.

```
export function compile(template: string, opts?: Options & { async: false }): TemplateFunction;
export function compile(template: string, opts: Options & { async: true }): AsyncTemplateFunction;
export function compile(template: string, opts: Options & { async?: never }): TemplateFunction;
export function compile(template: string, opts?: Options): TemplateFunction | AsyncTemplateFunction;
```

This both passes all the tests (yours and mine) and doesn't require conditional types.
The `boolean` case is also handled by the final case, so that's not necessary either.

Sorry for all the confusion with that giant conditional type; I agree it's not easy to read.